WT-9818 Disallow in_memory when using tiered storage

pmacko86 opened the pull request:
Disallow the use of the "in_memory" configuration to be used together with tiered storage.

The PR also removes the cleanup code from `__wt_tiered_conn_config`, because it resulted in double-free: On error, it would remove the current `bstorage`, but it would be removed later _again_ when `wiredtiger_open` cleans up, when it goes through the linked list of all block storages.

sueloverso reviewed (approved):
LGTM. Can you confirm that an error during `conn->reconfigure` of a tiered item also works correctly with your error path changes? (We might already have a tiered test that checks some reconfig errors. I haven't looked.)

keitharnoldsmith reviewed (approved):
LGTM

pmacko86 commented:
> LGTM. Can you confirm that an error during `conn->reconfigure` of a tiered item also works correctly with your error path changes? (We might already have a tiered test that checks some reconfig errors. I haven't looked.)

@sueloverso, `test_tiered04` tests reconfiguration—albeit just one case that is expected to succeed—which still works. I am not actually sure how to trigger the reconfiguration in `__wt_tiered_conn_config` to fail: If we are reconfiguring, it only updates the local retention number and the interval (using `__wt_config_gets`), so there is thus not any opportunity to fail. (And it shouldn't fail inside `__wt_config_gets` given that the configuration must have already passed its `__wt_config_check` by that point.)

FWIW, I added another test to `test_tiered21`: Since the `in_memory` option is not allowed in reconfigure, I added a check for that. This doesn't invoke the new error path, but I added it for completeness.
